pkg/cache/keyfunc.go

```diff
@@ -36,36 +79,37 @@ func DeletionHandlingMetaClusterNamespaceKeyFunc(obj interface{}) (string, error
 	return MetaClusterNamespaceKeyFunc(obj)
 }
 
+// ObjMetaClusterNamespaceKey is a convenient default KeyFunc which knows how to make
+// structured keys for API objects which implement meta.Interface.
+// This is the structured alternative to MetaClusterNamespaceKeyFunc;
+// putting such an object reference in a queue means that no parsing errors are possible downstream.
+func ObjMetaClusterNamespaceKey(obj interface{}) (Key, error) {
+	if key, ok := obj.(cache.ExplicitKey); ok {
+		return ParseKey(string(key))
+	}
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return Key{}, fmt.Errorf("object has no meta: %v", err)
+	}
+	return Key{Cluster: logicalcluster.From(meta), Namespace: meta.GetNamespace(), Name: meta.GetName()}, nil
+}
+
 // MetaClusterNamespaceKeyFunc is a convenient default KeyFunc which knows how to make
 // keys for API objects which implement meta.Interface.
 // The key uses the format <clusterName>|<namespace>/<name> unless <namespace> is empty, then
 // it's just <clusterName>|<name>, and if running in a single-cluster context where no explicit
 // cluster name is given, it's just <name>.
 func MetaClusterNamespaceKeyFunc(obj interface{}) (string, error) {
-	if key, ok := obj.(cache.ExplicitKey); ok {
-		return string(key), nil
-	}
-	meta, err := meta.Accessor(obj)
-	if err != nil {
-		return "", fmt.Errorf("object has no meta: %v", err)
-	}
-	return ToClusterAwareKey(logicalcluster.From(meta).String(), meta.GetNamespace(), meta.GetName()), nil
+	key, err := ObjMetaClusterNamespaceKey(obj)
+	return key.String(), err
 }
 
 // ToClusterAwareKey formats a cluster, namespace, and name as a key.
 func ToClusterAwareKey(cluster, namespace, name string) string {
```

Comment:
This has [downstream consumers](https://github.com/search?q=org%3Akcp-dev+-repo%3Akcp-dev%2Fapimachinery+kcpcache.ToClusterAwareKey&type=code) so we probably want to add a new variant. Folks use it for enumerating index values.

Reply:
I do not understand this comment.  This PR is not changing the signature or behavior of this function.  `NewKey` _is_ the new (structured) variant.

Reply:
Oops! I misread the diff and lost the `.String()` at the end of line 109. My apologies.